File: geomajas/face/geomajas-face-dojo/geomajas-dojo-client/src/main/resources/geomajas/widget/ScaleSelect.js
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

dojo.provide("geomajas.widget.ScaleSelect");
dojo.require("dijit.form.ComboBox");
dojo.requireLocalization("geomajas.widget", "scaleSelect");

/**
 * @fileoverview 
 * This widget connects to a MapWidget, and synchronizes with the map's scale.
 * It has a combobox, with which one can select a new scale level for the map,
 * or in which one can type a new scale level. But the synchronization goes in
 * both directions. If one zooms in or out on the map, this widget will 
 * automatically display the new scale.
 * If one tries to fill in a new scale level maually, be sure to type an 
 * integer: if you fill in integer X, the widget will apply scale 1:X.
 * 
 * Also, this widget can be added to a toolbar, with the toolbar.addChild() 
 * method.
 * 
 * TODO: A combobox is a ValidationTextBox, so perhaps we should take 
 * advantage of the regexp, the promptmessage, the invalidmessage, etc.
 * 
 * @author Pieter De Graef
 */
dojo.declare("geomajas.widget.ScaleSelect", [ dijit._Widget, dijit._Templated,
		dijit._Container ], {

	widgetsInTemplate :true,
	templatePath :dojo.moduleUrl("geomajas.widget", "html/ScaleSelect.html"),

	scaleText : "Scale : ", // The text that comes before the combobox.
	mapWidget : null, // Reference to the MapWidget with which to sync scales.

	/**
	 * Set the scale-text according to internationalization definitions.
	 */
	postMixInProperties : function (/*Object*/args, /*Object*/frag, /*Widget*/parent) {
		var widgetLocale = dojo.i18n.getLocalization("geomajas.widget", "scaleSelect");
		this.scaleText = widgetLocale.scaleText;
		this.inherited(arguments);
	},

	/**
	 * On startup of this widget, automatically set some scales to fill the
	 * combobox with. Also attach a function on the "onChange" event.
	 */
	startup : function() {
		var cmb = dijit.byId(this.id + ":cmb");
		if (cmb != null) {
			this.setScaleArray( [ 500, 1000, 2500, 5000, 7500, 10000, 20000,
					50000, 100000, 500000, 1000000, 5000000, 10000000 ]);
			cmb.onChange = dojo.hitch(this, "_onChange");
		}
		this.inherited(arguments);
	},

	/**
	 * This widget needs to be connected to a MapWidget to which it must
	 * synchronize scales. This function does that.
	 * @param mapWidget The MapWidget object to synchronize scales to.
	 */
	setMapWidget : function(mapWidget) {
		this.mapWidget = mapWidget;
		if (this.mapWidget != null) {
			this._updateScaleLabel(this.mapWidget.getMapView().getCurrentScale());
			this.connect(this.mapWidget.getMapView(), "setCurrentScale", dojo.hitch(this, "_updateScaleLabel"));
			this.connect(this.mapWidget.getScaleUtil(), "unitLengthCallback", dojo.hitch(this, "_onUnitLengthCallback"));
			
			var resolutions = this.mapWidget.getMapView().getResolutions();
			if (resolutions != null) {
				var scales = [];
				for(var i=resolutions.length-1; i>=0; i--) {
					var scale = parseFloat(1 / resolutions[i]);
					var temp = scale * this.mapWidget.getScaleUtil().getPixelWidth();
					var value = Math.round(this.mapWidget.getScaleUtil().getUnitLength() / temp);
					scales.push(value);
				}
				this.setScaleArray(scales);
			}
		}
	},

	/**
	 * Set a new list of scale options in the combobox.
	 * @param array The list of scale should be an array with integer values.
	 *              These will be filled in as (1:x) one by one, so better make
	 *              sure they are sorted! Example: [1000,5000,10000].
	 */
	setScaleArray : function(array) {
		if (array == null || !(array instanceof Array)) {
			log.debug("No array given in ScaleSelect.setScaleArray!");
			return;
		}
		var cmb = dijit.byId(this.id + ":cmb");
		if (cmb != null) {
			var store = {
				data : {
					identifier :"count",
					items : []
				}
			};
			for ( var i = 1; i <= array.length; i++) {
				var item = {
					name :"" + i,
					label :"1 : " + this._integerToDottedString(array[i - 1]),
					count :i
				};
				store.data.items.push(item);
			}
			cmb.store = new dojo.data.ItemFileReadStore(store);
		} else {
			log.error("Could not find the scale combobox. (" + this.id
					+ ":cmb)");
		}
	},



	//-------------------------------------------------------------------------
	// Private methods
	//-------------------------------------------------------------------------

	/**
	 * @private
	 * When a new value is entered in the combobox, try and zoom in or out on
	 * the map.
	 * @param label The label from the combobox. Either one of the selected 
	 *              values (1:x), or a simple integer (x).
	 */
	_onChange : function (label) {
		if (this.mapWidget == null) {
			log.error ("This ScaleSelect ("+this.id+") is not connected to a" +
					" MapWidget yet!");
			return;
		}
		if (label == null || label == "") {
			var scale = this.mapWidget.getMapView().getCurrentScale();
			this._updateScaleLabel(scale);
			return;
		}
		var scale = this.mapWidget.getScaleUtil().stringToScale(label);
		if (scale >= 0) {
			this.mapWidget.getMapView().setCurrentScale(scale);
		} else {
			var cmb = dijit.byId(this.id + ":cmb");
			cmb.setDisplayedValue("");
		}
	},

	/**
	 * @private
	 * React to the mapview's setCurrentScale, and update the combobox label,
	 * without having it fire the onChange event again.
	 * @param scale The current scale from the MapView object.
	 */
	_updateScaleLabel : function (scale) {
		if (this.mapWidget == null) {
			log.error ("This ScaleSelect ("+this.id+") is not connected to a" +
					" MapWidget yet!");
			return;
		}
		// Ask scale again, because it may have been snapped to a resolution:
		scale = this.mapWidget.getMapView().getCurrentScale();
		var label = this.mapWidget.getScaleUtil().scaleToString(scale);

		var cmb = dijit.byId(this.id + ":cmb");
		cmb._onChangeActive = false; // little hack.
		cmb.setDisplayedValue(label);
		cmb._onChangeActive = true;
	},

	/**
	 * @private
	 * Transforms an integer like 32000 to string 32.000
	 */
	_integerToDottedString : function (value) {
		var strValue = "" + value;
		var count = strValue.length - 3;
		while (count > 0) {
			var before = strValue.substring(0, count);
			var after = strValue.substring(count);
			strValue = before + "." + after;
			count = count - 3;
		}
		return strValue;
	},

	/**
	 * @private
	 */
	_onUnitLengthCallback : function (result) {
		var scale = this.mapWidget.getMapView().getCurrentScale();
		this._updateScaleLabel(scale);
	}
});

File: geomajas/face/geomajas-face-gwt/geomajas-gwt-client/src/main/java/org/geomajas/gwt/client/widget/ScaleSelect.java
/*
 * This file is part of Geomajas, a component framework for building
 * rich Internet applications (RIA) with sophisticated capabilities for the
 * display, analysis and management of geographic information.
 * It is a building block that allows developers to add maps
 * and other geographic data capabilities to their web applications.
 *
 * Copyright 2008-2010 Geosparc, http://www.geosparc.com, Belgium
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.geomajas.gwt.client.widget;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;

import org.geomajas.gwt.client.i18n.I18nProvider;
import org.geomajas.gwt.client.map.MapView;
import org.geomajas.gwt.client.map.event.MapViewChangedEvent;
import org.geomajas.gwt.client.map.event.MapViewChangedHandler;

import com.google.gwt.i18n.client.NumberFormat;
import com.smartgwt.client.widgets.Canvas;
import com.smartgwt.client.widgets.form.DynamicForm;
import com.smartgwt.client.widgets.form.fields.ComboBoxItem;
import com.smartgwt.client.widgets.form.fields.events.ChangedEvent;
import com.smartgwt.client.widgets.form.fields.events.ChangedHandler;
import com.smartgwt.client.widgets.form.fields.events.KeyPressEvent;
import com.smartgwt.client.widgets.form.fields.events.KeyPressHandler;
import com.smartgwt.client.widgets.form.validator.CustomValidator;

/**
 * A drop down selection box for setting and displaying the current scale of a map. The displayed scale is a relative
 * scale and is expressed as an abstract number, as opposed to the mapview scale which is expressed in pixels per map
 * unit. An arbitary scale can be filled in by the user. Implemented as a canvas so it can be added where needed.
 *
 * @author Jan De Moerloose
 */
public class ScaleSelect extends Canvas implements KeyPressHandler, ChangedHandler, MapViewChangedHandler {

	private MapView mapView;

	private DynamicForm form;

	private ComboBoxItem scaleItem;

	private static NumberFormat DENOMINATOR_FORMAT = NumberFormat.getFormat("###,###");

	// needed for item
	private List<String> valueList;

	// bidirectional lookup
	private LinkedHashMap<Double, String> scaleToValue;
	private LinkedHashMap<String, Double> valueToScale;

	// pixel length in map units
	private double pixelLength;

	/**
	 * Constructs a ScaleSelect that acts on the specified map view.
	 *
	 * @param mapView the map view
	 * @param pixelLength the pixel length in map units
	 */
	public ScaleSelect(MapView mapView, double pixelLength) {
		this.mapView = mapView;
		this.pixelLength = pixelLength;
		setWidth100();
		setHeight100();
		init();
	}

	/**
	 * Set the specified relative scale values in the combobox.
	 *
	 * @param scales array of relative scales (should be multiplied by pixelLength if in pix/m)
	 */
	public void setScales(Double... scales) {
		// sort decreasing
		Arrays.sort(scales, Collections.reverseOrder());
		// create lookup maps
		valueList = new ArrayList<String>();
		scaleToValue = new LinkedHashMap<Double, String>();
		valueToScale = new LinkedHashMap<String, Double>();
		for (int i = 0; i < scales.length; i++) {
			Double scale = scales[i];
			// eliminate doubles and null
			if (scale != null && !scaleToValue.containsKey(scale)) {
				String value = scaleToString(scale);
				valueList.add(value);
				scaleToValue.put(scale, value);
				valueToScale.put(value, scale);
			}
		}
		// set list
		scaleItem.setValueMap(valueList.toArray(new String[0]));
	}

	public MapView getMapView() {
		return mapView;
	}

	public void onMapViewChanged(MapViewChangedEvent event) {
		if (!event.isPanning()) {
			double currentScale = mapView.getCurrentScale() * pixelLength;
			setDisplayScale(currentScale);
		}
	}

	public void onKeyPress(KeyPressEvent event) {
		String name = event.getKeyName();
		if (name.equalsIgnoreCase("enter")) {
			reorderValues();
		}
	}

	public void onChanged(ChangedEvent event) {
		String value = (String) scaleItem.getValue();
		Double scale = valueToScale.get(value);
		if (scale != null) {
			mapView.setCurrentScale(scale / pixelLength, MapView.ZoomOption.LEVEL_CLOSEST);
		}
	}

	protected void setDisplayScale(double scale) {
		scaleItem.setValue(scaleToString(scale));
	}

	protected String scaleToString(double scale) {
		if (scale > 0 && scale < 1.0) {
			int denom = (int) Math.round(1. / scale);
			return "1 : " + DENOMINATOR_FORMAT.format(denom);
		} else if (scale >= 1.0) {
			int denom = (int) Math.round(scale);
			return DENOMINATOR_FORMAT.format(denom) + " : 1";
		} else {
			return "negative scale not allowed";
		}
	}

	protected Double stringToScale(String s) {
		String[] scale2 = s.split(":");
		if (scale2.length == 1) {
			return 1.0 / DENOMINATOR_FORMAT.parse(scale2[0].trim());
		} else {
			return DENOMINATOR_FORMAT.parse(scale2[0].trim()) / DENOMINATOR_FORMAT.parse(scale2[1].trim());
		}
	}

	private void init() {
		form = new DynamicForm();
		scaleItem = new ComboBoxItem();
		scaleItem.setTitle(I18nProvider.getToolbar().scaleSelect());
		scaleItem.setValidators(new ScaleValidator());
		scaleItem.setValidateOnChange(true);
		scaleItem.addKeyPressHandler(this);
		scaleItem.addChangedHandler(this);
		form.setFields(scaleItem);
		addChild(form);
		if (mapView.getResolutions() != null) {
			List<Double> scales = new ArrayList<Double>();
			for (Double resolution : mapView.getResolutions()) {
				scales.add(pixelLength / resolution);
			}
			setScales(scales.toArray(new Double[0]));
		}
		mapView.addMapViewChangedHandler(this);
	}

	private void reorderValues() {
		String value = (String) scaleItem.getValue();
		if (value != null) {
			Double scale = valueToScale.get(value);
			if (scale == null) {
				List<Double> newScales = new ArrayList<Double>(valueToScale.values());
				scale = stringToScale(value);
				newScales.add(scale);
				setScales(newScales.toArray(new Double[0]));
			}
			scaleItem.setValue(scaleToValue.get(scale));
			mapView.setCurrentScale(scale / pixelLength, MapView.ZoomOption.LEVEL_CLOSEST);
		}
	}

	/**
	 * Custom validation of user entered scale
	 */
	private class ScaleValidator extends CustomValidator {

		@Override
		protected boolean condition(Object value) {
			try {
				Double d = stringToScale((String) value);
				return d.doubleValue() >= 0;
			} catch (Throwable t) {
				return false;
			}
		}
	}
}
